x/aswap/handler.go

```diff
@@ -0,0 +1,307 @@
+package aswap
+
+import (
+	"bytes"
+	"crypto/sha256"
+	"time"
+
+	"github.com/iov-one/weave"
+	"github.com/iov-one/weave/coin"
+	"github.com/iov-one/weave/errors"
+	"github.com/iov-one/weave/migration"
+	"github.com/iov-one/weave/x"
+	"github.com/iov-one/weave/x/cash"
+	"github.com/tendermint/tendermint/libs/common"
+)
+
+const (
+	// pay swap cost up-front
+	createSwapCost  int64 = 300
+	returnSwapCost  int64 = 0
+	releaseSwapCost int64 = 0
+
+	// currently set to two days
+	day = time.Hour * 24
+	// amount of days for minTimeout
+	timeoutDays = 2
+	minTimeout  = timeoutDays * day
+
+	tagSwapId string = "swap-id"
+	tagAction string = "action"
+)
+
+// RegisterRoutes will instantiate and register
+// all handlers in this package
+func RegisterRoutes(r weave.Registry, auth x.Authenticator, cashctrl cash.Controller) {
+	bucket := NewBucket()
+
+	r.Handle(pathCreateSwap, migration.SchemaMigratingHandler("aswap", CreateSwapHandler{auth, bucket, cashctrl}))
+	r.Handle(pathReleaseSwap, migration.SchemaMigratingHandler("aswap", ReleaseSwapHandler{auth, bucket, cashctrl}))
+	r.Handle(pathReturnReturn, migration.SchemaMigratingHandler("aswap", ReturnSwapHandler{auth, bucket, cashctrl}))
+}
+
+// RegisterQuery will register this bucket as "/aswaps"
+func RegisterQuery(qr weave.QueryRouter) {
+	NewBucket().Register("aswaps", qr)
+}
+
+//---- create
+
+// CreateSwapHandler creates a swap
+type CreateSwapHandler struct {
+	auth   x.Authenticator
+	bucket Bucket
+	bank   cash.CoinMover
+}
+
+var _ weave.Handler = CreateSwapHandler{}
+
+// Check does the validation and sets the cost of the transaction
+func (h CreateSwapHandler) Check(ctx weave.Context, db weave.KVStore, tx weave.Tx) (*weave.CheckResult, error) {
+	_, err := h.validate(ctx, db, tx)
+	if err != nil {
+		return nil, err
+	}
+
+	res := &weave.CheckResult{
+		GasAllocated: createSwapCost,
+	}
+	return res, nil
+}
+
+// Deliver moves the tokens from sender to the swap account if all conditions are met.
+func (h CreateSwapHandler) Deliver(ctx weave.Context, db weave.KVStore, tx weave.Tx) (*weave.DeliverResult, error) {
+	msg, err := h.validate(ctx, db, tx)
+	if err != nil {
+		return nil, err
+	}
+
+	// create a swap object
+	swap := &Swap{
+		Metadata:     &weave.Metadata{},
+		Src:          msg.Src,
+		Recipient:    msg.Recipient,
+		Timeout:      msg.Timeout,
+		Memo:         msg.Memo,
+		PreimageHash: msg.PreimageHash,
+	}
+	obj, err := h.bucket.Build(db, swap)
+	if err != nil {
+		return nil, err
+	}
+
+	// deposit amounts
+	senderAddr := swap.Src
+	if err := moveCoins(db, h.bank, senderAddr, swap.Address, msg.Amount); err != nil {
```

Comment:
Swap.Address is nil??

Oh, wait it is set inside `bucket.Build`. But that smells like bad code. I mean, you return something and generally don't expect it to as a side-effect modify the arguments as well. If args are modified, it should be clear.

Why need the Address field anyway? And not just a method to calculate it?

Reply:
That was my thought too, but then Alex convinced me that this is going to be easier to query to determine the address. Not sure of the use-case here, but it makes sense. It's done in build because it's the only place we have both original swap and a generated id before it turns into orm object.

Reply:
I would prefer something more transparent.

Like 
```golang
swap.calculateAddress()
obj, err := Build(swap)
```

The function that calculates the address is more clearly able to modify it, then something "building" and returning an object. Still not sure about this Build idiom rather than just using Create